src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,380 @@
+        auto cursorArgs = winrt::make_self<CursorPositionEventArgs>();
+        _currentCursorPositionHandlers(*this, *cursorArgs);
+        const COORD cursorPos = { gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().X), gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().Y) };
+
+        // Get Font Info as we use this is the pixel size for characters in the display
+        auto fontArgs = winrt::make_self<FontInfoEventArgs>();
+        _currentFontInfoHandlers(*this, *fontArgs);
+
+        const float fontWidth = fontArgs->FontSize().Width;
+        const float fontHeight = fontArgs->FontSize().Height;
+
+        // Convert text buffer cursor position to client coordinate position within the window
+        COORD clientCursorPos;
+        COORD screenCursorPos;
+        THROW_IF_FAILED(ShortMult(cursorPos.X, gsl::narrow<SHORT>(fontWidth), &clientCursorPos.X));
+        THROW_IF_FAILED(ShortMult(cursorPos.Y, gsl::narrow<SHORT>(fontHeight), &clientCursorPos.Y));
+
+        // Convert from client coordinate to screen coordinate by adding window position
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.X, gsl::narrow_cast<SHORT>(windowBounds.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.Y, gsl::narrow_cast<SHORT>(windowBounds.Y), &screenCursorPos.Y));
+
+        // get any offset (margin + tabs, etc..) of the control within the window
+        const auto offsetPoint = this->TransformToVisual(nullptr).TransformPoint(winrt::Windows::Foundation::Point(0, 0));
+
+        // add the margin offsets if any
+        const auto currentMargin = this->Margin();
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.X, gsl::narrow_cast<SHORT>(offsetPoint.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.Y, gsl::narrow_cast<SHORT>(offsetPoint.Y), &screenCursorPos.Y));
+
+        // Get scale factor for view
+        const double scaleFactor = DisplayInformation::GetForCurrentView().RawPixelsPerViewPixel();
+
+        // Set the selection layout bounds
+        Rect selectionRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().TextBounds(ScaleRect(selectionRect, scaleFactor));
+
+        // Set the control bounds of the whole control
+        Rect controlRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().ControlBounds(ScaleRect(controlRect, scaleFactor));
+
+        // position textblock to cursor position
+        _canvas.SetLeft(_textBlock, clientCursorPos.X);
+        _canvas.SetTop(_textBlock, static_cast<double>(clientCursorPos.Y) + 2); // TODO GitHub #3642: Need a better way to align text in center
+
+        // width is cursor to end of canvas
+        _textBlock.Width(200); // TODO GitHub #3640: Determine proper Width
+        _textBlock.Height(fontHeight);
+
+        // TODO GitHub #3641: TSFInputControl FontSize needs to be larger than settings to match on screen font
+        _textBlock.FontSize(14);

```

Comment:
That's so very baffling.

Reply:
I'm still not 100% sure whether the height we get out of the terminal engine is in pixels or points, but if it's in pixels we should be able to DPI-transform it back to points... or ask the TerminalSettings hanging off the Control what the user's preferred point size was?

Reply:
I agree, my only explanation is XAML TextBlock vs. DirectWrite

Reply:
@DHowett-MSFT , figured it out, XAML TextBlock is expecting pixels (DIP), the font height was coming in as actual pixels.